### ISSUE TITLE
Batch normalization for RNNs

### ISSUE THREAD

**Issue by alquraishi** [tensorflow/tensorflow#1736]
I realize that this is hot off the press, but since batch normalization for feed forward layers is being added, any chance we can see it for RNNs? The performance improvements seem substantial.

http://arxiv.org/abs/1603.09025

**Comment by keveman** [comment]
@alquraishi, that definitely seems like would be a great addition to the existing RNN library in TF. Marking as 'contributions welcome', feel free to send a PR.

**Comment by mschonwe** [comment]
+1

**Comment by alquraishi** [comment]
FYI there's a torch implementation here: https://github.com/iassael/torch-bnlstm

**Comment by LeavesBreathe** [comment]
Yes I agree -- I'm trying to work on an implementation of it right now. If I get it to work, I'll submit a PR but the problem is that they are going to update LSTM soon so I don't know if it will conflict:

https://github.com/tensorflow/tensorflow/pull/2002

**Comment by JulesGM** [comment]
there is an implementation here of batch normalized lstm: https://github.com/OlavHN/bnlstm/blob/master/lstm.py
explained in the article http://olavnymoen.com/2016/07/07/rnn-batch-normalization
inspired from the paper https://arxiv.org/abs/1603.09025

**Comment by LinJM** [comment]
+1

**Comment by tgy** [comment]
That's also something I would be really interested in! :+1:

**Comment by carlthome** [comment]
I believe one of the important points made in Cooijmans paper is that you're supposed to collect mean/variances __per timestep__ during training (to let the LSTM state stabilize properly at the start of a new sequence), and then in test time you step through them (and if a test sequence is longer, you just keep with the last mean/var for the remainder of the sequence). OlavHN's implementation doesn't do this.

I have a version that includes this (but for ConvLSTMs, however it would be easy to copy+paste into FC-LSTMs): https://github.com/carlthome/tensorflow-convlstm-cell/blob/a647de6b86eff4c4bb502205d0ff9c1262869f19/ConvLSTMCell.py

**Comment by MichelHalmes** [comment]
Good day,

Is there any progress on this?

As pointed out by carlthome the implementation of @OlavHN does not compute distinct population statistics per timestep. To my understanding, @carlthome's implementation neither. To compute the timestep he uses:
```python
new_sequences = tf.placeholder(tf.bool)
...
increment_op = tf.cond(self._new_sequences,
                               lambda: tf.assign(timestep, 0),
                               lambda: tf.add(timestep, 1))
```
This basically means that `new_sequences` is constant per run and `timestep` will either be 0 or grow indefinitely.

I've tried to implement it myself but getting a variable to reflect the RNN-timestep in the `call()`  is tricky. Does anyone have idea on how to overcome this issue? 
Is there maybe a tensor from the outside one could access by name?
I'd be happy to take it from there...

Thanks in advance! :-)

**Comment by cooijmanstim** [comment]
If it helps, Sonnet has a TF implementation of a batch-normalized LSTM here: https://github.com/deepmind/sonnet/blob/master/sonnet/python/modules/gated_rnn.py

**Comment by Andres-Hernandez** [comment]
I will give it a try and make a push request by Monday

**Comment by Andres-Hernandez** [comment]
I have made the pull request
#14106

It follows the quoted paper, even in the initializer for the scaling variable, except that the statistics are taken "globally" rather than for each time step as recommended. The reason is that it uses tensorflow.python.layers.normalization.BatchNormalization to perform the actual normalization, and it would require that the variable creation/reuse be controlled from outside, and it seemed a strong reworking of it, so I decided to leave the pull request without any modification of BatchNormalization. If desired it could be implemented as a second step.

**Comment by alextp** [comment]
Closing as the pull request is closed.
